### ISSUE TITLE
generateInfoJson and outputKind "global"

### ISSUE THREAD

**Issue by achimmihca** [vojtechhabarta/typescript-generator#597]
### Scenario

Consider the project structure:

Root
| - Base
| - Leaf

where Leaf references Base. We are using the typescript-generator plugin already in Base.
Leaf holds different data classes. Thereby it also references classes from Base.

Example:
```
// In Base
public enum BaseEnum { YES, NO; }
public class BaseData { public String name; public BaseEnum baseEnum; }

// In Leaf
public class LeafData { public List<BaseData> baseDatas;}
```

So, for BaseData and BaseEnum there are already generated TS declarations.

### Problem

I wanted to do a second run of the typescript-generator plugin in Leaf.
Problem is that the interfaces and enums of Base are declared again. The second run knows nothing about the first run in Base.
TypeScript does not accept duplicate declarations of the enums (the duplicate interfaces would be ok because they are the same).

As workaround, I could name all already generated declarations in "customTypeMappings" configuration of the second run.
However, the first generated file has about 10000 lines and I would need to name some 1000 declarations.
Thus, this not a practical solution for me.

As I understand, [generateInfoJson](http://www.habarta.cz/typescript-generator/maven/typescript-generator-maven-plugin/generate-mojo.html#generateInfoJson) could solve this because it names declarations that already have been generated.
I would want to generate this file in the first run and reference it in the second run (and possibly further runs).

Sadly, we are not using a module system (yet) for out TS code, such that everything is in global scope.
At the moment, generateInfoJson can only be used with outputKind "module" but I need it with "global".

Would this be possible?
Thanks for the help!

**Comment by vojtechhabarta** [comment]
I think that since you have everything in one global scope you could just use typescript-generator output from Leaf module and ignore output from Base (or even not generate it at all). If you have more modules like Leaf1 and Leaf2 you could create new module that would depend on Leaf1 and Leaf2 and that module would generate complete output in one run. Or do you have some reason why to generate the output in multiple runs?

**Comment by achimmihca** [comment]
> Or do you have some reason why to generate the output in multiple runs?

The example is simplified.
In general (and in my case), the classes in Base and Leaf are different. Leaf uses (i.e. builds upon) a part of Base, but Base also provides different classes.
The goal is to generate a TS interface for all of them.

Our project structure is not optimal (it has grown historically).
Nonetheless, I think this feature would a good addition to typescript-generator.

**Comment by achimmihca** [comment]
I cloned the repo and tried to remove the check that forbids combination of `outputKind=global` and `generateInfoJson=true`.
Therefor, I change in Settings.java
`if (generateInfoJson && outputKind != TypeScriptOutputKind.module) {`
to 
`if (generateInfoJson && outputKind == TypeScriptOutputKind.ambientModule) {`

I found that the typescript-generator-info.json file is generated just fine then. I really think there is no need to forbid this combination.

I have only few issues left:
- I cannot make use of the generated info.json.
    - infoJson is a parameter of moduleDependencies. However, I do not use modules. I would need a new `infoJson` parameter
    - This new `infoJson` parameter should accept a list of info.json files. This would allow feeding it with output of multiple projects or a combination of manually created info.json file and automatically created info.json.
- The info.json does not include the custom type mappings that I configured in the pom.xml
    - As a result, one would have to repeat the same type mappings in following pom.xml files (redundant). I have defined some hundred custom type mappings and copying these to several projects is not a practical solution for me.
    - Instead I suggest to:
        - (a) also output custom type mappings from pom.xml to the info.json
            and/or
        - (b) enable the use of a list of files for `infoJson` parameter as mentioned above. This way, I could move the manually created type mappings to an infoJson file and reference this from multiple projects.

The suggested infoJson approach could be a more powerful replacement for the `customTypeMappings` parameter in pom.xml.
It allows composing multiple runs of the plugin and feed custom type information from several sources.

What do you think?

**Comment by vojtechhabarta** [comment]
It is surprising for me that you have so many custom type mapping. Why is that needed? Is it because those classes are handled during runtime specifically in Jackson by custom serializer? Or is it workaround for missing support for global scope infoJson?

How would you reference Base file from Leaf file? Using triple-slash directive `/// <reference path="*.d.ts" />`? Or leave it without  a reference and have just list of files in a project?

**Comment by achimmihca** [comment]
> How would you reference Base file from Leaf file?

The projects are compiled one after another.
We reference the generated d.ts files of previous projects in "include" property of tsconfig.json
This way, the compiler knows about the symbols of previous projects (remember that all these symbols are in global scope).

> It is surprising for me that you have so many custom type mapping. Why is that needed? 

Mostly two reasons for custom type mappings:
- (a) for manually created enums.
    - TypeScript complains about "declare const enum" when there is also such a normal enum. This is understandable.
- (b) for manually created classes where the type of a field does not (yet) match the type of the generated interface (e.g. `any` vs. concrete type).
    - This is due to the current migration from plain JS to TS. Once this has been completely finished, the generated interface should match the class such that the custom type mappings should not be needed anymore.

Also, there are few mappings with generics.

**Comment by achimmihca** [comment]
The question I have is: 
If I would implement the infoJson parameter as I need it, would this be merged? Or do you reject the concept?

Also I wonder: does the generated infoJson support generics? This would be nice for manually created info.json.

**Comment by vojtechhabarta** [comment]
I have some doubts about allowing dependencies on "global modules". There is a problem with potential conflicts when those modules are evolved in the future. Let's say there are modules `moduleA` and `moduleB` with following contents:
```
moduleA
  - Person

moduleB
  - Company
```

and module `moduleC` which depends on `moduleA` and `moduleB`. Everything works fine until developer of `moduleB` adds different `Person` class (which is perfectly fine from Java perspective). Then there is a conflict because this global type is defined in two modules. So this behavior somehow breaks module concept - adding a class should not break dependant modules.

But I agree that this feature can be useful despite the desribed problem. So I wanted to know what this feature exactly involves and implemented it in `global-module-dependencies` branch. Could you please check it? I extended existing `moduleDependencies` parameter with `global` property which needs to be `true` in this case and I also added the check for mentioned conflicts. I guess you will be able to find the details in my commit f2e9ad541c69f33d73ad786183cc16db0bfde224.

**Comment by achimmihca** [comment]
For documentation, I want to share the config in pom.xml:
```
<configuration>
...
    <!-- generate infoJson for this run -->
    <generateInfoJson>true</generateInfoJson>

    <!-- reference infoJson of previous run or manually created infoJson -->
    <moduleDependencies>
        <moduleDependency>
            <global>true</global>
            <infoJson>${project.basedir}/typings/CustomTypeMappings.json</infoJson>
        </moduleDependency>
    </moduleDependencies>
</configuration>
```

**Comment by achimmihca** [comment]
Also I created a Java file to transform the `<mapping>` entries to json format. Might be helpful for someone:
```
import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.StandardOpenOption;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Convert customTypeMappings to json.
 * Input must be one mapping per line in format <mapping>the.java.package.TheClass:TheTypeScriptSymbol</mapping>.
 * 
 * Execute via `java --source 11 "CustomTypeMappingsToJson.java" "path/to/CustomTypeMappings.txt"`
 */
public class CustomTypeMappingsToJson {

    private static final String DEFAULT_FILE_PATH = "./CustomTypeMappings.txt";
    private static final String OUTPUT_FILE_PATH = "./CustomTypeMappings.json";

    public static void main(String[] args) throws IOException {
        if (args.length < 1) {
            args = new String[] { DEFAULT_FILE_PATH };
        }

        File inputFile = new File(args[0]);
        if (!inputFile.exists()) {
            throw new RuntimeException("File does not exist: " + args[0]);
        }
        if (!inputFile.isFile()) {
            throw new RuntimeException("File-object is not a file: " + args[0]);
        }

        File outputFile = new File(OUTPUT_FILE_PATH);

        transformCustomTypeMappingsToJson(inputFile, outputFile);
    }

    private static void transformCustomTypeMappingsToJson(File inputFile, File outputFile) throws IOException {
        List<TypeMapping> typeMappings = new ArrayList<>();
        List<String> lines = Files.readAllLines(inputFile.getAbsoluteFile().toPath());
        for(int lineNumber = 0; lineNumber < lines.size(); lineNumber++)
        {
            try
            {
                String trimmedLine = lines.get(lineNumber).trim();
                if (trimmedLine.startsWith("<mapping>")
                    && trimmedLine.endsWith("</mapping>"))
                {
                    String mappingSeparatedByColon = trimmedLine
                    .replace("<mapping>", "")
                    .replace("</mapping>", "");
                    String[] javaClassAndTsSymbol = mappingSeparatedByColon.split(":");
                    TypeMapping typeMapping = new TypeMapping();
                    typeMapping.javaClass = javaClassAndTsSymbol[0];
                    typeMapping.tsSymbol = javaClassAndTsSymbol[1];
                    typeMappings.add(typeMapping);
                }
            }
            catch (Exception e)
            {
                System.err.println("Error at line " + lineNumber + ": " + e.getMessage());
                e.printStackTrace();
            }
        }

        StringBuilder sb = new StringBuilder();
        appendLine(sb, "{");
        appendLine(sb, "  \"classes\": [");
        appendLine(sb, typeMappings.stream()
            .map(typeMapping -> typeMapping.toJson("    "))
            .collect(Collectors.joining(",\n")));
        appendLine(sb, "  ]");
        appendLine(sb, "}");

        if (outputFile.exists())
        {
            outputFile.delete();
        }
        Files.writeString(outputFile.getAbsoluteFile().toPath(), sb.toString(), StandardOpenOption.CREATE);
    }

    private static void appendLine(StringBuilder sb, String text)
    {
        sb.append(text);
        sb.append("\n");
    }

    private static class TypeMapping {
        public String javaClass;
        public String tsSymbol;

        public String toJson(String indentation) {
            return indentation + "{\n"
                + indentation + "  \"javaClass\": \"" + javaClass + "\",\n"
                + indentation + "  \"typeName\": \"" + tsSymbol + "\"\n"
                + indentation + "}";
        }
    }
}
```

**Comment by vojtechhabarta** [comment]
Released in [2.30.840](https://github.com/vojtechhabarta/typescript-generator/releases/tag/v2.30.840).
